pytest/tests/sanity/rpc_finality.py

```diff
@@ -1,66 +1,108 @@
+
+        utils.wait_for_blocks(nodes[0], target=3)
 
-        acc0_balance = int(nodes[0].get_account('test0')['result']['amount'])
-        acc1_balance = int(nodes[0].get_account('test1')['result']['amount'])
+        balances = {
+            account: int(nodes[0].get_account('test0')['result']['amount'])
+            for account in ['test0', 'test1']
+        }
 
         token_transfer = 10
         latest_block_hash = nodes[0].get_latest_block().hash_bytes
         tx = sign_payment_tx(nodes[0].signer_key, 'test1', token_transfer, 1,
                              latest_block_hash)
         logger.info("About to send payment")
-        logger.info(nodes[0].send_tx_and_wait(tx, timeout=200))
+        # this transaction will be added to the block (probably around block 5)
+        # and the the receipts & transfers will happen in the next block (block 6).
+        # This function should return as soon as block 6 arrives in node0.
+        logger.info(nodes[0].send_tx_and_wait(tx, timeout=10))
         logger.info("Done")
 
-        # wait for doomslug finality
-        time.sleep(5)
-        for i in range(2):
-            acc_id = 'test0' if i == 0 else 'test1'
-            acc_no_finality = nodes[0].get_account(acc_id)
-            acc_doomslug_finality = nodes[0].get_account(acc_id, "near-final")
-            acc_nfg_finality = nodes[0].get_account(acc_id, "final")
-            if i == 0:
-                self.assertEqual(int(acc_no_finality['result']['amount']),
-                                 acc0_balance - token_transfer)
-                self.assertEqual(int(acc_doomslug_finality['result']['amount']),
-                                 acc0_balance - token_transfer)
-                self.assertEqual(int(acc_nfg_finality['result']['amount']),
-                                 acc0_balance - token_transfer)
+        # kill one validating node so that block cannot be finalized.
+        nodes[2].kill()
+
+        print(
+            f"Block height is {nodes[0].get_latest_block().height} (should be 6)"
+        )
+
+        # So now the situation is following:
+        # Block 6 (head) - has the final receipt (that adds state to test1)
+        # Block 5 (doomslug) - has the transaction (so this is the moment when state is removed from test0)
+        # Block 4 (final) - has no information about the transaction.
+
+        # So with optimistic finality: test0 = -10, test1 = +10
+        # with doomslug (state as of block 5): test0 = -10, test1 = 0
+        # with final (state as of block 4): test0 = 0, test1 = 0
+
+        for acc_id in ['test0', 'test1']:
+            amounts = [
+                int(nodes[0].get_account(acc_id, finality)['result']['amount'])
+                - balances[acc_id]
+                for finality in ["optimistic", "near-final", "final"]
+            ]
+            print(f"Account amounts: {acc_id}: {amounts}")
+
+            if acc_id == 'test0':
+                self.assertEqual([-10, -10, 0], amounts)
```

Comment:
Maybe we should check that balances are just lower/higher than original balances instead of checking exact values, due to our work on changes with refund processing? @anthony-near 
Though we may not want to change behaviour of such exposed API.

Reply:
The changes to refund receipts have been reverted and are unlikely to be re-added.

Reply:
So exact values should work for tests.